src/java.base/share/classes/sun/nio/fs/AbstractWatchKey.java

```diff
@@ -28,16 +28,39 @@
 import java.nio.file.*;
 import java.util.*;
 
+import jdk.internal.util.ArraysSupport;
+import sun.security.action.GetPropertyAction;
+
 /**
  * Base implementation class for watch keys.
  */
 
 abstract class AbstractWatchKey implements WatchKey {
 
+    private static final int DEFAULT_MAX_EVENT_LIST_SIZE = 512;
+
     /**
-     * Maximum size of event list (in the future this may be tunable)
+     * Maximum size of event list before dropping events and signalling OVERFLOW
      */
-    static final int MAX_EVENT_LIST_SIZE    = 512;
+    @SuppressWarnings("removal")
+    static final int MAX_EVENT_LIST_SIZE;
+    static {
+        String rawValue = GetPropertyAction.privilegedGetProperty(
+            "jdk.nio.file.maxWatchEvents",
+            String.valueOf(DEFAULT_MAX_EVENT_LIST_SIZE));
+        int intValue;
+        try {
+            // Clamp to max array length to signal OVERFLOW and drop events
+            // before OOMing.
+            intValue = Math.clamp(
+                Long.decode(rawValue),
+                DEFAULT_MAX_EVENT_LIST_SIZE,
```

Comment:
Since 0 or -1 seem like pretty clear intent to "disable" watching as far as possible, I went with a clamp to 1 in that case. But I'm open to making that a hard error instead if you prefer that.

Reply:
I think clamping to 1 is fine. An error hear might be awkward.